src/protocol/vm/state_builder.rs

```diff
@@ -0,0 +1,461 @@
+            hasher.update(method_name.as_bytes());
+            let result = hasher.finalize();
+            result[..4].to_vec()
+        };
+
+        let to_address = decoded
+            .split(':')
+            .nth(1)
+            .ok_or_else(|| {
+                SimulationError::DecodingError("Invalid decoded string format".into())
+            })?;
+
+        let timestamp = Utc::now()
+            .naive_utc()
+            .and_utc()
+            .timestamp() as u64;
+
+        let parsed_address: rAddress = to_address
+            .parse()
+            .map_err(|_| SimulationError::DecodingError("Invalid address format".into()))?;
+
+        let sim_params = SimulationParameters {
+            data: selector.to_vec().into(),
+            to: parsed_address,
+            block_number: self.block.number,
+            timestamp,
+            overrides: Some(HashMap::new()),
+            caller: *EXTERNAL_ACCOUNT,
+            value: 0.into(),
+            gas_limit: None,
+        };
+
+        let sim_result = engine
+            .simulate(&sim_params)
+            .map_err(SimulationError::SimulationEngineError)?;
+
+        let address = decode(&[ParamType::Address], &sim_result.result)
+            .map_err(|_| SimulationError::DecodingError("Failed to decode ABI".into()))?
+            .into_iter()
+            .next()
+            .ok_or_else(|| {
+                SimulationError::DecodingError(
+                    "Couldn't retrieve address from simulation for stateless contracts".into(),
+                )
+            })?;
+
+        address
+            .to_string()
+            .parse()
+            .map_err(|_| SimulationError::DecodingError("Couldn't parse address to string".into()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use super::*;
+    use ethers::types::{H160, H256};
+
+    #[test]
+    fn test_build_without_required_fields() {
+        let id = "pool_1".to_string();
+        let tokens = vec![H160::zero()];
+        let block = BlockHeader { number: 1, hash: H256::default(), timestamp: 234 };
+
+        let result = tokio_test::block_on(VMPoolStateBuilder::new(id, tokens, block).build());
+
+        assert!(result.is_err());
+        match result.unwrap_err() {
+            SimulationError::NotInitialized(field) => assert_eq!(field, "Adapter contract"),
+            _ => panic!("Unexpected error type"),
+        }
+    }
+
+    #[test]
+    fn test_engine_setup() {
+        let id = "pool_1".to_string();
+        let token2 = H160::from_str("0000000000000000000000000000000000000002").unwrap();
+        let token3 = H160::from_str("0000000000000000000000000000000000000003").unwrap();
+        let tokens = vec![token2, token3];
+        let block = BlockHeader { number: 1, hash: H256::default(), timestamp: 234 };
+
+        let mut builder = VMPoolStateBuilder::new(id, tokens, block);
+
+        let engine = tokio_test::block_on(builder.get_default_engine()).unwrap();
```

Comment:
It's to avoid having a full runtime running during the test. I read that it's better to use this if it's a simple unit test that calls an async function.

But I'm not 100% sure if it's really worth having a dev dependency for it or not, maybe we don't care about the overhead of a runtime here.